File: testing-framework/backend/src/services/outlineToFreemind.ts
function escapeMm(text: string): string {
  if (!text) return "";
  return text
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

export function outlineTextToFreemind(outlineText: string): string {
  const lines = outlineText.split(/\r?\n/).filter((l) => l.trim() !== "");
  if (lines.length === 0) return "";
  const first = lines[0].replace(/^\t+/, "").trim();
  if (!first) return "";
  const rest = lines.slice(1);
  const stack: { level: number }[] = [];
  const result: string[] = [];
  for (const line of rest) {
    const stripped = line.replace(/^\t+/, "");
    const level = line.length - stripped.length;
    while (stack.length > 0 && stack[stack.length - 1].level >= level) {
      stack.pop();
      result.push("</node>");
    }
    result.push('<node TEXT="' + escapeMm(stripped.trim()) + '">');
    stack.push({ level });
  }
  while (stack.length > 0) {
    stack.pop();
    result.push("</node>");
  }
  const childrenXml = result.join("\n");
  return `<?xml version="1.0" encoding="UTF-8"?>
<map version="1.0.1">
<node TEXT="${escapeMm(first)}">
${childrenXml}
</node>
</map>`;
}

File: testing-framework/backend_python/app/services/outline_to_freemind.py
import re


def escape_mm(text: str) -> str:
    if not text:
        return ""
    return (
        text.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
        .replace("'", "&apos;")
    )


def outline_text_to_freemind(outline_text: str) -> str:
    lines = [l for l in re.split(r"\r?\n", outline_text) if l.strip() != ""]
    if not lines:
        return ""
    first = re.sub(r"^\t+", "", lines[0]).strip()
    if not first:
        return ""
    rest = lines[1:]
    stack: list[int] = []
    result: list[str] = []
    for line in rest:
        stripped = re.sub(r"^\t+", "", line)
        level = len(line) - len(stripped)
        while stack and stack[-1] >= level:
            stack.pop()
            result.append("</node>")
        result.append('<node TEXT="' + escape_mm(stripped.strip()) + '">')
        stack.append(level)
    while stack:
        stack.pop()
        result.append("</node>")
    children_xml = "\n".join(result)
    return f"""<?xml version="1.0" encoding="UTF-8"?>
<map version="1.0.1">
<node TEXT="{escape_mm(first)}">
{children_xml}
</node>
</map>"""
